File: java/src/topo_sort/randomTestData.java
package topo_sort;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

class randomTestData {
	private static final Random rand = new Random(0);

	List<List<Integer>> graph;
	int[][] adj;
	boolean hasCycle;

	randomTestData() {
		var n = 1 + rand.nextInt(100);
		graph = new ArrayList<List<Integer>>();
		for (int i = 0; i < n; i++) {
			graph.add(new ArrayList<Integer>());
		}

		var m = 1 + rand.nextInt(n*2);
		var ins = new int[n];
		for (int i = 0; i < m; i++) {
			var u = rand.nextInt(n);
			var v = rand.nextInt(n);
			graph.get(u).add(v);
			ins[v]++;
		}

		adj = new int[n][];
		for (int u = 0; u < n; u++) {
			adj[u] = new int[graph.get(u).size()];
			for (int i = 0; i < adj[u].length; i++) {
				adj[u][i] = graph.get(u).get(i);
			}
		}

		var q = new int[n];
		var r = 0;
		for (int v = 0; v < n; v++) {
			if (ins[v] == 0) {
				q[r++] = v;
			}
		}

		for (int l = 0; l < r; l++) {
			for (var v: adj[q[l]]) {
				if (--ins[v] == 0) {
					q[r++] = v;
				}
			}
		}

		hasCycle = r < n;
	}
}

File: go/topo_sort/random_test_data.go
package topo_sort

import (
	"math/rand/v2"
)

type randomTestData struct {
	graph, adj [][]int
	hasCycle   bool
}

var random = rand.New(rand.NewPCG(0, 0))

func newRandomTestData() *randomTestData {
	var td randomTestData

	n := 1 + random.IntN(100)
	td.graph = make([][]int, n)

	m := 1 + random.IntN(n*2)
	ins := make([]int, n)
	for range m {
		u := random.IntN(n)
		v := random.IntN(n)
		td.graph[u] = append(td.graph[u], v)
		ins[v]++
	}

	td.adj = make([][]int, n)
	for u := range n {
		td.adj[u] = make([]int, len(td.graph[u]))
		for i := 0; i < len(td.adj[u]); i++ {
			td.adj[u][i] = td.graph[u][i]
		}
	}

	q := make([]int, n)
	r := 0
	for v := range n {
		if ins[v] == 0 {
			q[r] = v
			r++
		}
	}

	for l := 0; l < r; l++ {
		for _, v := range td.adj[q[l]] {
			if ins[v]--; ins[v] == 0 {
				q[r] = v
				r++
			}
		}
	}

	td.hasCycle = r < n

	return &td
}
